File: 441.排列硬币.py
#
# @lc app=leetcode.cn id=441 lang=python
#
# [441] 排列硬币
#
# https://leetcode-cn.com/problems/arranging-coins/description/
#
# algorithms
# Easy (40.49%)
# Likes:    61
# Dislikes: 0
# Total Accepted:    22.6K
# Total Submissions: 55.2K
# Testcase Example:  '5'
#
# 你总共有 n 枚硬币，你需要将它们摆成一个阶梯形状，第 k 行就必须正好有 k 枚硬币。
#
# 给定一个数字 n，找出可形成完整阶梯行的总行数。
#
# n 是一个非负整数，并且在32位有符号整型的范围内。
#
# 示例 1:
#
#
# n = 5
#
# 硬币可排列成以下几行:
# ¤
# ¤ ¤
# ¤ ¤
#
# 因为第三行不完整，所以返回2.
#
#
# 示例 2:
#
#
# n = 8
#
# 硬币可排列成以下几行:
# ¤
# ¤ ¤
# ¤ ¤ ¤
# ¤ ¤
#
# 因为第四行不完整，所以返回3.
#
#
#

# @lc code=start


class Solution(object):
    def arrangeCoins(self, n):
        """
        :type n: int
        :rtype: int
        """
        # 常规
        # i = 0
        # total = 0
        # while True:
        #     if total + i + 1 > n:
        #         break
        #     i += 1
        #     total = total + i
        # return i

        # 数学
        # i = 0
        # while i*(i+1)/2 <= n:
        #     i += 1
        # return i - 1

        # 二分查找
        # 找到第一个大于n的左边
        if n <= 1:
            return n
        l = 1
        r = n
        while l < r:
            m = (l+r) // 2
            if self.calc(m) > n:
                r = m
            else:
                l = m+1
        return l-1

    def calc(self, k):
        return (1+k) * k / 2

# @lc code=end

File: 441.排列硬币.go
/*
 * @lc app=leetcode.cn id=441 lang=golang
 *
 * [441] 排列硬币
 *
 * https://leetcode.cn/problems/arranging-coins/description/
 *
 * algorithms
 * Easy (45.19%)
 * Likes:    282
 * Dislikes: 0
 * Total Accepted:    123.9K
 * Total Submissions: 274.2K
 * Testcase Example:  '5'
 *
 * 你总共有 n 枚硬币，并计划将它们按阶梯状排列。对于一个由 k 行组成的阶梯，其第 i 行必须正好有 i 枚硬币。阶梯的最后一行 可能 是不完整的。
 *
 * 给你一个数字 n ，计算并返回可形成 完整阶梯行 的总行数。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：n = 5
 * 输出：2
 * 解释：因为第三行不完整，所以返回 2 。
 *
 *
 * 示例 2：
 *
 *
 * 输入：n = 8
 * 输出：3
 * 解释：因为第四行不完整，所以返回 3 。
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1 <= n <= 2^31 - 1
 *
 *
 */
package main

// @lc code=start
/*
1. 为什么 l=mid 而不是 l = mid+1, 因为mid可能就是结果, +1会造成错过结果;
2. 为什么 mid := (l + r + 1) / 2 而不是 (l + r) / 2, 考虑到1的前提和 (0,1)的情况,不+1会造成死循环.
*/
func arrangeCoins(n int) int {
	l, r := 1, n
	for l < r {
		mid := (l + r + 1) / 2
		need := mid * (mid + 1)
		if need <= n*2 {
			l = mid
		} else {
			r = mid - 1
		}
	}
	return l
}

// @lc code=end
